### PR TITLE
Make tree compilation *optional* when parsing sdf/urdf files

### PR DESCRIPTION
**Pull request by SeanCurtis-TRI**
This enables being able to programmatically modify collision filter groups
of parsed files. It does this by providing an optional flag that defers
tree compilation (which used to be a defacto step in parsing).

1. URDF and SDF parsers are updated to propagate the flags.
2. Testing is provided to confirm behavior.
3. Incidental corrections to providing collision filter group support in
   sdf parsing.
   a. The doxygen module on the subject has been updated.
   b. SDF specific semantics have been documented and tested.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8297)

### CODE REVIEW

**Comment by SeanCurtis-TRI** [comment]
+@rcory for feature review, please.

---

Review status: 0 of 11 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-:-L71jIwtgsl15ZAj9OrH:b-yvq8xq)*

**Comment by rcory** [comment]
:lgtm: with a few BTWs.

One general comment based on our f2f discussion: I'm wondering if it would be cleaner to have a single version of each `AddModelInstance...` function that takes in an optional `do_compile` boolean as its last argument. I realize that according to the style guide, function outputs are supposed to come last in the argument list, but perhaps we make an exception in this case? The upside being we get a cleaner interface.

---

Reviewed 11 of 11 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[multibody/collision/collision_filter_doxygen.h, line 204 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L75pQVL8OrXgpvPZEQy:-L75pQVL8OrXgpvPZEQz:beruj15) ([raw file](https://github.com/robotlocomotion/drake/blob/906d3ea432c550700c26b9c73b1f245c99b035ae/multibody/collision/collision_filter_doxygen.h#L204)):*
> ```objc
> one model to the first group, and all elements of the second model to the second
> group.  Set either group to _ignore_ the other (or both, redundancy doesn't
> hurt).
> ```

BTW, based on our f2f conversation, perhaps add a note here that this also applies to filtering between different models in an sdf.

---

*[multibody/collision/collision_filter_doxygen.h, line 252 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L75CGOl2OffCIZnvIZY:-L75CGOm3SeNfTE4h_wW:b-c4fvl3) ([raw file](https://github.com/robotlocomotion/drake/blob/906d3ea432c550700c26b9c73b1f245c99b035ae/multibody/collision/collision_filter_doxygen.h#L252)):*
> ```objc
> `<cfg>` tags are children of each `<model>` tag. The `<cfg>` tags can only
> include links that are defined *in that model* in its membership lists. The
> implication of that is collision filter groups defined in sdf file contain
> ```

BTW, ...groups defined in sdf file**s**... (should be plural)

---

*[multibody/collision/collision_filter_doxygen.h, line 255 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L75qMFc3wsjCFECqR_b:-L75qMFc3wsjCFECqR_c:b-fm23yy) ([raw file](https://github.com/robotlocomotion/drake/blob/906d3ea432c550700c26b9c73b1f245c99b035ae/multibody/collision/collision_filter_doxygen.h#L255)):*
> ```objc
> bodies from a single model only. However, a collision filter group in *one*
> model can ignore a collision filter group in another model. See
> @ref /drake/multibody/collision/test/multi_model_groups.sdf for an example.
> ```

BTW, for consistency it would be great to include a snippet here as well, similar to what you have above for urdf.

---

*[multibody/collision/collision_filter_doxygen.h, line 260 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L75rbhW6K-4-ILswv5k:-L75rbhW6K-4-ILswv5l:b-itjpt8) ([raw file](https://github.com/robotlocomotion/drake/blob/906d3ea432c550700c26b9c73b1f245c99b035ae/multibody/collision/collision_filter_doxygen.h#L260)):*
> ```objc
> 
> In addition to parsing the collision filter groups from the urdf/sdf files,
> there is an API on the RigidBodyTree that allows manipulation. However, the API
> ```

BTW, to be consistent with the wording in the paragraph below, perhaps reword this as: "there is a legacy API on the ..."

---

*[multibody/collision/collision_filter_doxygen.h, line 275 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L75s5KO23GoU1Qhb--G:-L75s5KO23GoU1Qhb--H:bpdcf9q) ([raw file](https://github.com/robotlocomotion/drake/blob/906d3ea432c550700c26b9c73b1f245c99b035ae/multibody/collision/collision_filter_doxygen.h#L275)):*
> ```objc
> altered.
> 
> There is a _second_ parsing interface that allows for suppressing the automatic
> ```

BTW, "parsing interface"-->"API"? Something about reading "API", "legacy API", and "parsing interface" confused me a bit.

Perhaps a rewording like: "There is a _second_ API that allows for collision filter groups to be programmatically altered by suppressing the automatic tree compilation".

---

*[multibody/collision/collision_filter_doxygen.h, line 282 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L75u0KW6pvgjlophNSB:-L75u0KW6pvgjlophNSC:bflebhk) ([raw file](https://github.com/robotlocomotion/drake/blob/906d3ea432c550700c26b9c73b1f245c99b035ae/multibody/collision/collision_filter_doxygen.h#L282)):*
> ```objc
> responsibility to make sure that RigidBodyTree::compile() is invoked before
> doing any work on the tree.
> ```

BTW, can we point to `collision_filter_group_test.cc` here as an example of this programmatic version?

---

*[multibody/parsers/sdf_parser.h, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L75zFzc-0uy3_Xyu7AN:-L75zFzc-0uy3_Xyu7AO:b-mjh85s) ([raw file](https://github.com/robotlocomotion/drake/blob/906d3ea432c550700c26b9c73b1f245c99b035ae/multibody/parsers/sdf_parser.h#L30)):*
> ```objc
>  *
>  * There are two versions: one in which the tree is automatically "compiled"
>  * after a successful parse and one compilation depends on an input parameter,
> ```

BTW, wording. "... and one in which the compilation depends...".

Same goes for similar comment headers defined here and `urdf_parser.h`. The wording just seems a little off.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-:-L76NmNL3NuSvKHiQ-ot:bm3wcnr)*

**Comment by rcory** [comment]
+@soonho-tri for platform review please.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-:-L76PEIBA0Qu_KHDMJfv:b-vt9vbo)*

**Comment by SeanCurtis-TRI** [comment]
Most comments addressed. PTAL

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[multibody/collision/collision_filter_doxygen.h, line 204 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L75pQVL8OrXgpvPZEQy:-L76sxGYSkf0nch_HwI6:benzku8) ([raw file](https://github.com/robotlocomotion/drake/blob/906d3ea432c550700c26b9c73b1f245c99b035ae/multibody/collision/collision_filter_doxygen.h#L204)):*
<details><summary><i>Previously, rcory (Rick Cory) wrote…</i></summary><blockquote>

BTW, based on our f2f conversation, perhaps add a note here that this also applies to filtering between different models in an sdf.
</blockquote></details>

I think I came away with a different thought from our f2f.

I'm *not* in favor of mentioning the sdf details here. That belongs in the the in-file specification discussion below.

The message I came away with is that you were surprised that collisions were filtered by unilateral declarations (e.g., A ignores B, but B says nothing about A).  And this paragraph clearly documents this. 

---

*[multibody/collision/collision_filter_doxygen.h, line 252 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L75CGOl2OffCIZnvIZY:-L76tOwS9tKonA9WWmxO:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/906d3ea432c550700c26b9c73b1f245c99b035ae/multibody/collision/collision_filter_doxygen.h#L252)):*
<details><summary><i>Previously, rcory (Rick Cory) wrote…</i></summary><blockquote>

BTW, ...groups defined in sdf file**s**... (should be plural)
</blockquote></details>

Done

---

*[multibody/collision/collision_filter_doxygen.h, line 255 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L75qMFc3wsjCFECqR_b:-L76tS-aoRPjAjjaJT_p:b-rlw99l) ([raw file](https://github.com/robotlocomotion/drake/blob/906d3ea432c550700c26b9c73b1f245c99b035ae/multibody/collision/collision_filter_doxygen.h#L255)):*
<details><summary><i>Previously, rcory (Rick Cory) wrote…</i></summary><blockquote>

BTW, for consistency it would be great to include a snippet here as well, similar to what you have above for urdf.
</blockquote></details>

I'd argue that a clickable link to the actual file would be the real solution -- after all, it's a concrete, running example and it doesn't need to be duplicated. However, it's not *currently* a clickable link. So, that's the part I should figure out.

---

*[multibody/collision/collision_filter_doxygen.h, line 260 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L75rbhW6K-4-ILswv5k:-L76tzyKmNaYkm6HF0_9:b6200uf) ([raw file](https://github.com/robotlocomotion/drake/blob/906d3ea432c550700c26b9c73b1f245c99b035ae/multibody/collision/collision_filter_doxygen.h#L260)):*
<details><summary><i>Previously, rcory (Rick Cory) wrote…</i></summary><blockquote>

BTW, to be consistent with the wording in the paragraph below, perhaps reword this as: "there is a legacy API on the ..."
</blockquote></details>

Done

"legacy" here doesn't refer to the API for configuring collision filter groups -- it refers to the parsing API. I've attempted to clarify this. PTAL.

---

*[multibody/collision/collision_filter_doxygen.h, line 275 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L75s5KO23GoU1Qhb--G:-L76uESYyvMNIwyWaJ7O:b318yep) ([raw file](https://github.com/robotlocomotion/drake/blob/906d3ea432c550700c26b9c73b1f245c99b035ae/multibody/collision/collision_filter_doxygen.h#L275)):*
<details><summary><i>Previously, rcory (Rick Cory) wrote…</i></summary><blockquote>

BTW, "parsing interface"-->"API"? Something about reading "API", "legacy API", and "parsing interface" confused me a bit.

Perhaps a rewording like: "There is a _second_ API that allows for collision filter groups to be programmatically altered by suppressing the automatic tree compilation".
</blockquote></details>

Done  interface swapped for API

---

*[multibody/collision/collision_filter_doxygen.h, line 282 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L75u0KW6pvgjlophNSB:-L76uaM3nN5Qk0rTmRDZ:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/906d3ea432c550700c26b9c73b1f245c99b035ae/multibody/collision/collision_filter_doxygen.h#L282)):*
<details><summary><i>Previously, rcory (Rick Cory) wrote…</i></summary><blockquote>

BTW, can we point to `collision_filter_group_test.cc` here as an example of this programmatic version?
</blockquote></details>

Done

---

*[multibody/parsers/sdf_parser.h, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L75zFzc-0uy3_Xyu7AN:-L76vJrNWVej_cPWo5sV:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/906d3ea432c550700c26b9c73b1f245c99b035ae/multibody/parsers/sdf_parser.h#L30)):*
<details><summary><i>Previously, rcory (Rick Cory) wrote…</i></summary><blockquote>

BTW, wording. "... and one in which the compilation depends...".

Same goes for similar comment headers defined here and `urdf_parser.h`. The wording just seems a little off.
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-:-L76vLTyaCXClPwrQynt:b-u7iolc)*

**Comment by SeanCurtis-TRI** [comment]
Review status: 9 of 12 files reviewed at latest revision, 2 unresolved discussions.

---

*[multibody/collision/collision_filter_doxygen.h, line 255 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L75qMFc3wsjCFECqR_b:-L76yCCsM4nFehN_AYvM:b6b99xb) ([raw file](https://github.com/robotlocomotion/drake/blob/906d3ea432c550700c26b9c73b1f245c99b035ae/multibody/collision/collision_filter_doxygen.h#L255)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I'd argue that a clickable link to the actual file would be the real solution -- after all, it's a concrete, running example and it doesn't need to be duplicated. However, it's not *currently* a clickable link. So, that's the part I should figure out.
</blockquote></details>

I've made it a clickable link.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8297)*

**Comment by soonho-tri** [comment]
:lgtm: with a few cosmetic BTWs.

---

Reviewed 7 of 11 files at r1, 1 of 1 files at r2, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L79EukALfT0y5xKiz9b:-L79EukALfT0y5xKiz9c:bvexkhw) (no related file):*
We've had a problem with an invalidated XML file (#8270). So I've validated the new files (`.sdf`/`.urdf`) introduced by this PR via `xmllint` and found a few errors. Some of them might be because of our own extensions, but it'd be great if you can confirm it. Could you check?

```
xmllint --noout --schema ./urdf.xsd ./collision/test/filter_groups_in_file.urdf
./urdf.xsd:155: element element: Schemas parser error : Element '{http://www.w3.org/2001/XMLSchema}element': Invalid value for maxOccurs (must be 0 or 1).
./urdf.xsd:157: element element: Schemas parser error : Element '{http://www.w3.org/2001/XMLSchema}element': Invalid value for maxOccurs (must be 0 or 1).
```

```
xmllint --noout --schema ./root.xsd ./collision/test/filter_groups_in_file.sdf
./collision/test/filter_groups_in_file.sdf:10: element collision: Schemas validity error : Element 'collision': The attribute 'name' is required but missing.
./collision/test/filter_groups_in_file.sdf:22: element collision: Schemas validity error : Element 'collision': The attribute 'name' is required but missing.
./collision/test/filter_groups_in_file.sdf:34: element collision: Schemas validity error : Element 'collision': The attribute 'name' is required but missing.
./collision/test/filter_groups_in_file.sdf:46: element collision: Schemas validity error : Element 'collision': The attribute 'name' is required but missing.
./collision/test/filter_groups_in_file.sdf:76: element collision_filter_group: Schemas validity error : Element 'collision_filter_group': This element is not expected.
```

To test it, you need to download the SDF/URDF schema files:

 - http://sdformat.org/schemas/root.xsd
 - https://raw.githubusercontent.com/ros/urdfdom/master/xsd/urdf.xsd

FYI, `xmllint` is a part of `libxml2-utils` in Xenial. Using homebrew (macOS), you need to install `xmlstarlet` package.

---

*[bindings/pydrake/multibody/rigid_body_tree_py.cc, line 246 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L76SQL8QoM-mntGSufK:-L76SQL8QoM-mntGSufL:b-tcb7yj) ([raw file](https://github.com/robotlocomotion/drake/blob/62ae6a483a8476ca65a7ced243382c058e0bfed7/bindings/pydrake/multibody/rigid_body_tree_py.cc#L246)):*
> ```cpp
>       "AddModelInstanceFromUrdfStringSearchingInRosPackages",
>       py::overload_cast<
>           const std::string&, const drake::parsers::PackageMap&,
> ```

FYI, META,

1) `drake::` part is not necessary since we're already in it.
2) Consider adding `using std::string/shared_ptr` to make things less verbose.

---

*[multibody/collision/test/collision_filter_group_test.cc, line 508 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L79IYtpRF9m8B1N3HGH:-L79IYtpRF9m8B1N3HGI:bqv69h3) ([raw file](https://github.com/robotlocomotion/drake/blob/03c113bc3726ad49e516e16fd6a234ebfb926a52/multibody/collision/test/collision_filter_group_test.cc#L508)):*
> ```cpp
> 
> // These tests read the files `filter_groups_in_file.[sdf|urdf]`. Each file
> // represents an identical tree with built in collisions where all collisions
> ```

BTW, nit, `built in collisions` -> `built-in collisions`

---

*[multibody/collision/test/multi_model_groups.sdf, line 47 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L79Ge-YS1-S_pRqtK1q:-L79Ge-YS1-S_pRqtK1r:b73rd5a) ([raw file](https://github.com/robotlocomotion/drake/blob/03c113bc3726ad49e516e16fd6a234ebfb926a52/multibody/collision/test/multi_model_groups.sdf#L47)):*
> ```
>         </collision_filter_group>
>     </model>
> </sdf>
> ```

BTW, need a newline here?

---

*[multibody/collision/test/no_filter_groups.sdf, line 34 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L76SJdyii6yOeSYjhhq:-L76SJdyii6yOeSYjhhr:b73rd5a) ([raw file](https://github.com/robotlocomotion/drake/blob/62ae6a483a8476ca65a7ced243382c058e0bfed7/multibody/collision/test/no_filter_groups.sdf#L34)):*
> ```
>         </link>
>     </model>
> </sdf>
> ```

BTW, need a newline here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-:-L79JAearIPXixIUwa6z:bspz0zl)*

**Comment by SeanCurtis-TRI** [comment]
Review status: 7 of 12 files reviewed at latest revision, 7 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L79EukALfT0y5xKiz9b:-L7Av0oaEgaOCgK99hvf:btx5aga) (no related file):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

We've had a problem with an invalidated XML file (#8270). So I've validated the new files (`.sdf`/`.urdf`) introduced by this PR via `xmllint` and found a few errors. Some of them might be because of our own extensions, but it'd be great if you can confirm it. Could you check?

```
xmllint --noout --schema ./urdf.xsd ./collision/test/filter_groups_in_file.urdf
./urdf.xsd:155: element element: Schemas parser error : Element '{http://www.w3.org/2001/XMLSchema}element': Invalid value for maxOccurs (must be 0 or 1).
./urdf.xsd:157: element element: Schemas parser error : Element '{http://www.w3.org/2001/XMLSchema}element': Invalid value for maxOccurs (must be 0 or 1).
```

```
xmllint --noout --schema ./root.xsd ./collision/test/filter_groups_in_file.sdf
./collision/test/filter_groups_in_file.sdf:10: element collision: Schemas validity error : Element 'collision': The attribute 'name' is required but missing.
./collision/test/filter_groups_in_file.sdf:22: element collision: Schemas validity error : Element 'collision': The attribute 'name' is required but missing.
./collision/test/filter_groups_in_file.sdf:34: element collision: Schemas validity error : Element 'collision': The attribute 'name' is required but missing.
./collision/test/filter_groups_in_file.sdf:46: element collision: Schemas validity error : Element 'collision': The attribute 'name' is required but missing.
./collision/test/filter_groups_in_file.sdf:76: element collision_filter_group: Schemas validity error : Element 'collision_filter_group': This element is not expected.
```

To test it, you need to download the SDF/URDF schema files:

 - http://sdformat.org/schemas/root.xsd
 - https://raw.githubusercontent.com/ros/urdfdom/master/xsd/urdf.xsd

FYI, `xmllint` is a part of `libxml2-utils` in Xenial. Using homebrew (macOS), you need to install `xmlstarlet` package.

</blockquote></details>

Great comment!  I've modified my workflow.

There is *one* major problem here, however.  The `urdf.xsd` uses `maxOccurs='unbounded'` in numerous locations. This is causing xmllint to fail (it feels that `maxOccurs` should be limited to a value of 0 or 1. This is definitely an error in `xmllint`. Strangely, the `sdf.xsd` file has a *single* reference to `maxOccurs='unbounded'`. I'm not sure why that gets through.

For now, I'm going to use the tool to the extent that I can. It can validate sdf to a certain degree, but not urdfs. The best we can say is that we can validate that it's valid XML, even if it doesn't interact with the schema properly.

---

*[bindings/pydrake/multibody/rigid_body_tree_py.cc, line 246 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L76SQL8QoM-mntGSufK:-L7B1h7e4nWkSO6x31r-:bm2rgka) ([raw file](https://github.com/robotlocomotion/drake/blob/62ae6a483a8476ca65a7ced243382c058e0bfed7/bindings/pydrake/multibody/rigid_body_tree_py.cc#L246)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, META,

1) `drake::` part is not necessary since we're already in it.
2) Consider adding `using std::string/shared_ptr` to make things less verbose.
</blockquote></details>

Done Excellent point. I took it a step further and, this time, noted that there were already `using` declarations for those types anyways.  :)

---

*[multibody/collision/test/collision_filter_group_test.cc, line 508 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L79IYtpRF9m8B1N3HGH:-L7B1r-sVFOvMWAd531M:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/03c113bc3726ad49e516e16fd6a234ebfb926a52/multibody/collision/test/collision_filter_group_test.cc#L508)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, nit, `built in collisions` -> `built-in collisions`
</blockquote></details>

Done

---

*[multibody/collision/test/multi_model_groups.sdf, line 47 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L79Ge-YS1-S_pRqtK1q:-L7B1t-k6MAPp5LJcHt-:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/03c113bc3726ad49e516e16fd6a234ebfb926a52/multibody/collision/test/multi_model_groups.sdf#L47)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, need a newline here?
</blockquote></details>

Done

---

*[multibody/collision/test/no_filter_groups.sdf, line 34 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L76SJdyii6yOeSYjhhq:-L7B1vIz2Y91hahPteRB:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/62ae6a483a8476ca65a7ced243382c058e0bfed7/multibody/collision/test/no_filter_groups.sdf#L34)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, need a newline here?
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8297)*

**Comment by soonho-tri** [comment]
Reviewed 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/8297#-L79EukALfT0y5xKiz9b:-L7BOm2Hn4156TufzzBE:b-x7kv41) (no related file):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Great comment!  I've modified my workflow.

There is *one* major problem here, however.  The `urdf.xsd` uses `maxOccurs='unbounded'` in numerous locations. This is causing xmllint to fail (it feels that `maxOccurs` should be limited to a value of 0 or 1. This is definitely an error in `xmllint`. Strangely, the `sdf.xsd` file has a *single* reference to `maxOccurs='unbounded'`. I'm not sure why that gets through.

For now, I'm going to use the tool to the extent that I can. It can validate sdf to a certain degree, but not urdfs. The best we can say is that we can validate that it's valid XML, even if it doesn't interact with the schema properly.
</blockquote></details>

Thanks for the update!

I just ran the new tests on my mac and they all passed. So I think we should be fine. I expect that we will have a systematic solution as https://github.com/RobotLocomotion/drake/issues/8271 is worked out.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8297)*
